### ISSUE TITLE
Unbounded pattern cache

### ISSUE THREAD

**Issue by mridgway** [ReactTraining/react-router#3875]
## Version
4.x

## Steps to reproduce

``<Link to={`${pathname}/${id}`} />``

You can see the cache grow by recursively clicking names in the [Recursive Paths example](https://react-router-website-uxmsaeusnn.now.sh/recursive-paths).

```js
Object.keys(cache);
// ["/", "/quick-start", "/basic", "/url-parameters", "/auth-workflow", "/preventing-transitions", "/no-match-handling", "/query-params", "/recursive-paths", "/sidebar", "/animated-transitions", "/ambiguous-matches", "/route-config", "/BrowserRouter", "/HashRouter", "/MemoryRouter", "/ServerRouter", "/Match", "/Miss", "/Link", "/NavigationPrompt", "/Redirect", "/:id", "/2/:id", "/2/1/:id", "/2/1/3/:id", "/2/1/3/2/:id", "/2/1/3/2/3/:id", "/2/1/3/2/3/2/:id", "/2/1/3/2/3/2/3/:id", "/2/1/3/2/3/2/3/2/:id", "/2/1/3/2/3/2/3/2/3/:id", "/2/1/3/2/3/2/3/2/3/2/:id", "/2/1/3/2/3/2/3/2/3/2/1/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/1/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/1/0/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/1/0/3/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/1/0/3/2/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/1/0/3/2/3/:id", "/2/1/3/2/3/2/3/2/3/2/1/0/1/0/2/1/0/1/:id"]
```

## Behavior

The dynamic creation of routes using relative `to` in the `Link` component has potential to cache an infinitely large number of regular expressions in https://github.com/ReactTraining/react-router/blob/v4/modules/matchPattern.js#L11. This could be exploited on server-rendered applications by accessing several variations of routes and overflowing the server's memory.

An LRU cache with a finite size could patch this, but finding a way to not have to create new regular expressions for each path extension (match only on the relative path rather than the full path somehow?) would be ideal.

**Comment by timdorr** [comment]
Hmm, an interesting point. Want to submit a PR to add the LRU cache instead of the simple object cache?

**Comment by alisd23** [comment]
I'll take a stab at this

**Comment by alisd23** [comment]
Wouldn't any scenario where this could happen be nonsensical though? I can't think of any situation where a relative link with a url parameter, which can be called recursively, would **not** be a bug.

So given that, is it still worth guarding against this anyway? Or should the dev be responsible for avoiding situations like this?

**Comment by jedwards1211** [comment]
@alisd23 what about routes that map to folders in a directory tree, like in dropbox?  Though I'm not sure using a url parameter would be the best way to model that.

**Comment by alisd23** [comment]
@jedwards1211 Actually looking at the code, even patterns that miss are cached - so there is always an infinite number of possible patterns that can be cached. I can see how that could cause memory trouble on the **client** (it would just take a lot of matches), but I can't really see how that could cause problems with server rendering.

But yeah, for the risk of overflowing memory on the client alone, it's probably worth switching to LRU cache.

**Comment by ryanflorence** [comment]
fixed in #4004
